### DISCUSSION TITLE
Declaration of Constants Data within eKuiper

### DISCUSSION THREAD

**Discussion by BlancoMY** [lf-edge/ekuiper#3197] [Q&A] [answered]
Hello eKuiper community! 👋
I'm currently working on integrating eKuiper into our project and I'm exploring the possibility of declaring persisted constants data directly within eKuiper, without relying on external sources. I would like to use these constants within the rules (SQL & graph rules) for streamlined processing.
Is there a recommended approach or feature within eKuiper that facilitates the declaration and utilization of constants data internally? I'd appreciate any insights, examples, or best practices regarding this scenario.

Thank you all for your help!

**Comment by ngjaying** [comment, marked as answer]
You can define a key/value in the sqlite file data/extState.db, table keyed_state. Then in your rule, you can access by [get-keyed-state](https://ekuiper.org/docs/en/latest/sqls/functions/other_functions.html#) function.

**Reply by BlancoMY** [reply]
Thank you that's very helpful
